app/oracle.py

```diff
@@ -165,53 +181,82 @@ def prompt(prompt_message, prompt_end):
             continue
 
 
+def get_previous_frame():
+    """Since the contract lacks a method that returns the time of last report and the reported numbers
+    we are using web3.py filtering to fetch it from the contract events."""
+    logging.info('Getting previously reported numbers (will be fetched from events)...')
+    SECONDS_PER_ETH1_BLOCK = 14
+    latest_block = w3.eth.getBlock('latest')
+    #Calculate earliest block to limit scanning depth
+    from_block = int((latest_block['timestamp']-genesis_time)/SECONDS_PER_ETH1_BLOCK)
+    events = oracle.events.Completed.getLogs(fromBlock=from_block, toBlock='latest')
+    if events:
+        event = events[-1]
+        epoch = event['args']['epochId']
+        balance = event['args']['beaconBalance']
+        beacon_validators = event['args']['beaconValidators']
+        timestamp = w3.eth.getBlock(event['blockHash'])['timestamp']
+        return (epoch, balance, beacon_validators, timestamp)
+    else:
+        logging.warning('No events on the contract')
+        return False
+
 logging.info('Starting the main loop')
 while True:
-
-    # Get the frame and validators keys from ETH1 side
+    # Get previously reported data 
+    prev_frame = get_previous_frame()
+    if prev_frame:
+        prev_epoch, prev_balance, prev_validators, prev_timestamp = prev_frame
+        logging.info(f'Previously reported epoch: {prev_epoch}')
+        logging.info(f'Previously reported beaconBalance: {prev_balance} wei or {prev_balance/1e18} ETH')
+        logging.info(f'Previously reported beaconValidators: {prev_validators}')
+        logging.info(f'Timestamp of previous report: {datetime.datetime.fromtimestamp(prev_timestamp)} or {prev_timestamp}')
+
+    # Get the the epoch that is both finalized and reportable
     current_frame = oracle.functions.getCurrentFrame().call()
-    reportable_epoch = current_frame[0]
-    logging.info(f'Reportable epoch: {reportable_epoch}')
+    potentially_reportable_epoch = current_frame[0]
+    logging.info(f'Potentially reportable epoch: {potentially_reportable_epoch} (from ETH1 contract)')
+    finalized_epoch_beacon = beacon.get_finalized_epoch()
+    logging.info(f'Last finalized epoch: {finalized_epoch_beacon} (from Beacon)')
+    epoch = min(potentially_reportable_epoch, (finalized_epoch_beacon // epochs_per_frame) * epochs_per_frame)
+    slot = epoch * slots_per_epoch
+    logging.info(f'Reportable state: epoch:{epoch} slot:{slot}')
+
     validators_keys = get_validators_keys(registry)
     logging.info(f'Total validator keys in registry: {len(validators_keys)}')
+    balance, validators = beacon.get_balances(slot, validators_keys)
+
+    logging.info(f'Lido validators\' sum. balance on Beacon: {balance} wei or {balance/1e18} ETH')
+    logging.info(f'Lido validators visible on Beacon: {validators}')
+
+    if prev_frame:
+        balance_change = balance - prev_balance
```

Comment:
use the same logic as in the smart contract: balance change is not the important metric, reward/slashing size is and it's calculated in a way more complex than this
reward should be shown in absolute and relative metrics, along with an APR